invenio_drafts_resources/services/draft.py

```diff
@@ -71,85 +62,138 @@ def draft_cls(self):
-        pid, record = self.resolve_draft(id_)
-        self.require_permission(identity, "read", record=record)
+        pid, draft = self.pid_manager.resolve(id_, draft=True)
+        self.require_permission(identity, "read", record=draft)
         # Todo: how do we deal with tombstone pages
-        return self.resource_unit(pid=pid, record=record)
+        return self.resource_unit(pid=pid, record=draft, links=None)
+
+    def update_draft(self, id_, data, identity):
+        """Replace a record."""
+        # TODO: etag and versioning
+        pid, draft = self.pid_manager.resolve(id_, draft=True)
+        # Permissions
+        self.require_permission(identity, "update", record=draft)
+        validated_data = self.data_validator.validate(data, partial=True)
+        # FIXME: extract somewhere else
+        self._patch_data(draft, validated_data)
+        draft.update(draft.dumps())
+        draft.commit()
+
+        if self.indexer:
+            self.indexer.index(draft)
+
+        return self.resource_unit(pid=pid, record=draft, links=None)
 
     def create(self, data, identity):
         """Create a draft for a new record.
 
         It does not eagerly create the associated record.
         """
         self.require_permission(identity, "create")
-        validated_data = self.draft_data_validator.validate(data)
+        validated_data = self.data_validator.validate(data, partial=True)
         rec_uuid = uuid.uuid4()
-        pid = self.minter(record_uuid=rec_uuid, data=validated_data)
+        pid = self.pid_manager.mint(record_uuid=rec_uuid, data=validated_data)
         draft = self.draft_cls.create(validated_data, id_=rec_uuid)
 
         db.session.commit()  # Persist DB
         if self.indexer:
             self.indexer.index(draft)
 
-        return self.config.resource_unit_cls(pid=pid, record=draft)
+        return self.resource_unit(pid=pid, record=draft, links=None)
+
+    def _patch_data(self, record, data):
+        """Temporarily here until the merge strategy is set."""
+        # TODO: Implement better update/merge strategy
+        for key in data.keys():
+            record[key] = data[key]
 
     def edit(self, id_, data, identity):
-        """Create a draft for an existing record.
+        """Create a new revision or a draft for an existing record.
 
         :param id_: record PID value.
         """
-        pid, record = self.resolve(id_)
+        pid, record = self.pid_manager.resolve(id_)
         self.require_permission(identity, "create")
-        validated_data = self.draft_data_validator.validate(data)
-        draft = self.draft_cls.create(validated_data, id_=record.id,
+        validated_data = self.data_validator.validate(data, partial=True)
+        self._patch_data(record, validated_data)
+        draft = self.draft_cls.create(record.dumps(), id_=record.id,
                                       fork_version_id=record.revision_id)
         db.session.commit()  # Persist DB
```

Comment:
Question: Above in `update_draft` we do `draft.commit()`, shouldn't we do the same here?

Reply:
Since the delete and create do something like `db.session.delete(record.model)`, you need to `db.session.commit()` the session itself. The operation just marks the SQLAlchemy model for deletion, it's persisted when the session is commited.

The `record.commit` is basically a necessary evil because SQLAlchemy cannot automatically  track changes to JSON (blob) columns. So it's just merging on it. Maybe @slint (who gave me this explanation) or @lnielsen can give more details.

My understanding is:
- Delete and Create sort of "execute plain delete and insert statements" (under the hood obviously). Therefore, they are full actions that need to be just commited. They can mark the full entity (self.model)
- To update we have to nest + merge, so we can "run an specific update statement on a blob".  Mainly because it  cannot track the `self.model.json` (cuz it is a blob (?)).  So we need to merge that specific field instead of update + commit session. 

I think the confusion might be due to naming. In the end the `record.commit` is only called on updates... so it might as well be part of said function?

Reply:
Example given by @slint 
```python
from sqlalchemy import flag_modified

class Record(db.Model):

    id = db.Column(db.UUID, primary_key=True, default=uuid.uuid4)
    json = db.Column(db.JSON)
    title = db.Column(db.Str)


# Just creates an SQLAlchemy model object
record = Record(json={'foo': 'bar'}, title='some title')

# adds the model to the session (you can find it in `db.session.new`)
db.session.add(record)

# Commits the current session state to the DB. SQLAlchemy checks what things
# you have in `db.session.{new,dirty,deleted}` and then generates the appropriate
# SQL statements. e.g. now it will send:
#    INSERT INTO records (json, title) VALUES ({'foo': 'bar'}, 'some title');
db.session.commit()

# Since we set directly the field, there is a "__set__" method called on the "title" member
# which marks the object as "dirty" and adds it to the session. After this you can
# check in `db.session.dirty` and you'll find the `record` object
record.title = 'something else'
# In the same fashion as above, this will generate:
#    UPDATE records WHERE id = '<record-uuid>' SET title = 'something else'
db.session.commit()


# This will not trigger the `json.__set__` since we're modifying the dictionary
# value directly... So if you check in `db.session.dirty` you won't find the `record` model
record.json['foo'] = 'baz'
# If instead we did something like:
# 
#    record.json = {'foo': 'baz'}
# 
# ...it would go through `json.__set__` and thus mark the object as "dirty"

# To achieve what we want we flag the object as dirty manually (this is what `record.commit()` does in Invenio):
flag_modified(record)

# This will send the update as expected:
#     UPDATE records WHERE id = '...' SET json = {'foo': 'baz'}
db.session.commit()
```

Reply:
Thank you! That is quite helpful in understanding some of the idiosyncracies of SQLAlchemy.